### DISCUSSION TITLE
I little help needed here.

### DISCUSSION THREAD

**Discussion by eduuh** [cajhin/capsicain#64] [Q&A]
I have been using this custom keyboard, 

![Vyikd](https://user-images.githubusercontent.com/31909722/145934877-6eb96218-8b93-453f-bbaa-f76e9c4bfa89.png)

Finally I was able to  partially create the mapping using the software. Some of the combination, did not seem to work on Capsicain so I remaped those key with Ahk for the mean time. The problem with ahk, some of this mapping don't get sent when am working in PowerShell.

Here my  Capsicain.ini

```
GLOBAL ActiveConfigOnStartup 1
GLOBAL StartAHK

[CONFIG_1]
INCLUDE LAYOUT_COLEMAK

REWIRE Caps LSHF   
REWIRE h  RET
REWIRE RET RSHF 

REWIRE RSHF RCtrl 
REWIRE LSHF LCtrl 

[LAYOUT_COLEMAK]
ALPHA_FROM  q  w  e  r  t     u  i  o  p      a  s  d  f  g     j  k  l  ;     z  x  c  v / . ,  n  m  [ ' \ y b  
ALPHA_TO    q  w  f  p  g     j  l  u  y      a  r  s  t  d     k  n  e  i     x  c  b  v . , h  z  m  ; o = nop nop 
ALPHA_END 
```

I need to remapping the entire number row with, to define the symbols  needed on those keys. Some problem faced is that i could not do remap let say 1 to ! directly using rewire or the Alpha. It would error. .

![Capture](https://user-images.githubusercontent.com/31909722/145936247-aab2d422-7f7c-444d-abd4-3a78d085f1bd.PNG)

How can I be able to remap this to capsicain.ini ,can i add more scancode to the codebase? 

```
]:::
`::?
1::!
2::[
3::]
4::"
5::{
6::}
7::'
8::)
9::(
0::\
=::|

LAlt & n::Send, {1}
LAlt & e::Send, {2}
LAlt & i::Send, {3}
LAlt & o::Send, {0}
LAlt & l::Send, {4}
LAlt & y::Send, {6}
LAlt & u::Send, {5}
LAlt & '::7
LAlt & )::Send, {8}
LAlt & (::Send, {9}
LAlt & j::Send, {+}
LAlt & h::Send, {#}
LAlt & z::Send, {$}
LAlt & m::Send, {ASC 037}
LAlt & k::Send, {ASC 045}


RAlt & a::Send, {Backspace}
RAlt & r::Send, {Esc}
RAlt & s::Send, {Home}
RAlt & t::Send, {End}
RAlt & d::Send, {Delete}

RAlt & i::Send, {Right}
RAlt & e::Send, {Down}
RAlt & n::Send, {Left}
RAlt & u::Send, {Up}
RAlt & o::Send, {Enter}

RAlt & q::Send, {@}
RAlt & w::Send, {~}
RAlt & f::Send, {<}
RAlt & p::Send, {>}
RAlt & g::Send, {{}

RAlt & x::Send, {*}
RAlt & c::Send, {`}
RAlt & b::Send, {&}
RAlt & v::Send, {|}
```

Thanks alot for the assistance.  

Thanks @cajhin for this awesome software. I don't have to mess up with registry keys anymore. This is by far the simplest custom keyboard set up i have created on windows. Thanks alot.

**Comment by cajhin** [comment]
Hi Edwin

first, that's a nice and clear email trouble report.
Is there a tool that generates the visual layout below or did you paint it manually?


About your issue, you need to understand that capsicain deals with keys. Only keys.
AHK does a lot of things, keys characters windows signals kitchen sink.

Now, your mapping [1] -> [!] is doomed, because there is no [!] key on regular keyboards.
How to create the character depends on the keyboard layout that is set in Windows. For English, the "!" CHARACTER is [Shift]+[1]

So, the mapping 1 -> !  is:

COMBO 1 [ ] > moddedKey (1 + ...&)  #!

("When key 1 is pressed together with [] any modifier, send LSHIFT+[1] ")

Same goes for the other characters that are made with a key combo.
Beware that those combos are layout dependent, so if Windows switches from "English" to "German" or "EN-International", keys may break. For example, quotes are shift+2 in German but Shift+' in English.

For other labels like ESC LEFT HOME see 
https://github.com/cajhin/capsicain/wiki/How-to-find-the-Key-Labels



And about your layout, nobody asked but obviously I like to dabble with layouts :)
Your virtual numberpad is non-standard. If you are still learning it, I would prefer to shift the top row 7 8 9 to the left. When your fingers have learned that, you can also type on a regular laptop numpad. If you shift everything down a row, you'll also be at home on a real numpad.
Moving the shift keys up is a nice idea, haven't seen that. I work around the frequent stressful pinky stretches by moving the shifts inside (left/right). Works for me, because I use my meta key on capslock more than the shift keys.
I don't like the number row very much, it's hard to hit the keys consistently. I'd try to at least move ? and ! to an easier position.
) and ( position is a mistake I hope :)

Good luck learning that
j
